File: experiments/016-LanguageComparisonRepeated/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260917-061134-48859/analysis/acceptance-project/src/main/java/simshotels/Floor.java
package simshotels;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.function.Predicate;
import java.util.function.Supplier;
import java.util.function.ToIntFunction;
import java.util.stream.IntStream;

public class Floor {

    private final ArrayList<Room> rooms;
    private final HashMap<String, Integer> prices;

    // instance creation (class side)

    public static Floor withRoomsAndPrices(Number aNumberOfRooms, HashMap<String, Integer> aPriceList) {
        Floor.assertIsValidNumberOfRooms(aNumberOfRooms);
        Floor.assertIsValidPriceList(aPriceList);

        return new Floor(aNumberOfRooms.intValue(), aPriceList);
    }

    // initialization

    private Floor(int aNumberOfRooms, HashMap<String, Integer> aPriceList) {
        rooms = new ArrayList<>();
        IntStream.rangeClosed(1, aNumberOfRooms).forEach(ix -> rooms.add(new Room()));
        prices = new HashMap<>(aPriceList);
    }

    // testing

    public boolean isAvailable() {
        return rooms.stream().allMatch(Room::isAvailable);
    }

    // guests

    public void receiveAtRoom(String aGuestType, int aRoomNumber) {
        roomNumbered(aRoomNumber).receive(aGuestType);
    }

    public void receiveWithReservationAtRoom(String aGuestType, int aRoomNumber) {
        roomNumbered(aRoomNumber).receiveWithReservation(aGuestType);
    }

    public void reserveRoom(int aRoomNumber) {
        roomNumbered(aRoomNumber).reserve();
    }

    private Room roomNumbered(int aRoomNumber) {
        if (aRoomNumber < 1 || aRoomNumber > totalRooms()) {
            Floor.signalRoomNumberDoesNotExist();
        }
        return rooms.get(aRoomNumber - 1);
    }

    // accounting

    public int totalLosses() {
        return sumOverRooms(room -> room.lossUsingIfAbsentGuestType(prices, Floor::unknownGuestTypeBlock));
    }

    public int totalProfits() {
        return sumOverRooms(room -> room.profitUsingIfAbsentGuestType(prices, Floor::unknownGuestTypeBlock));
    }

    private int sumOverRooms(ToIntFunction<Room> aRoomValue) {
        return rooms.stream().mapToInt(aRoomValue).sum();
    }

    private static Integer unknownGuestTypeBlock() {
        Floor.signalUnknownGuestType();
        return null;
    }

    // totals

    public int totalRooms() {
        return rooms.size();
    }

    public int totalRoomsAvailable() {
        return countRoomsThat(Room::isAvailable);
    }

    public int totalRoomsOccupied() {
        return countRoomsThat(Room::isOccupied);
    }

    public int totalRoomsReserved() {
        return countRoomsThat(Room::isReserved);
    }

    private int countRoomsThat(Predicate<Room> aCondition) {
        return (int) rooms.stream().filter(aCondition).count();
    }

    // assertions (class side)

    public static void assertIsValidNumberOfRooms(Number aNumberOfRooms) {
        Floor.assertIsPositiveIfFalse(aNumberOfRooms, () -> Floor.signalNumberOfRoomsMustBePositive());
        Floor.assertIsIntegerIfFalse(aNumberOfRooms, () -> Floor.signalNumberOfRoomsMustBeInteger());
    }

    public static void assertIsValidPriceList(HashMap<String, Integer> aPriceList) {
        if (aPriceList.isEmpty()) {
            Floor.signalNoPrices();
        }
        aPriceList.values().forEach(price ->
            Floor.assertIsPositiveIfFalse(price, () -> Floor.signalPriceMustBePositive()));
    }

    public static void assertIsIntegerIfFalse(Number aNumber, Runnable signalsBlock) {
        if (!(aNumber instanceof Integer)) {
            signalsBlock.run();
        }
    }

    public static void assertIsPositiveIfFalse(Number aNumber, Runnable signalsBlock) {
        if (!(aNumber.doubleValue() > 0)) {
            signalsBlock.run();
        }
    }

    // error messages (class side)

    public static String numberOfRoomsMustBeIntegerErrorDescription() {
        return "Number of rooms must be integer";
    }

    public static String numberOfRoomsMustBePositiveErrorDescription() {
        return "The number of rooms must be positive";
    }

    public static String priceMustBeIntegerErrorDescription() {
        return "Price must be integer";
    }

    public static String priceMustBePositiveErrorDescription() {
        return "Price must be positive";
    }

    public static String pricesListCannotBeEmptyErrorDescription() {
        return "Prices List must not be empty";
    }

    public static String roomNumberDoesNotExistErrorDescription() {
        return "Room number does not exist";
    }

    public static String unknownGuestTypeErrorDescription() {
        return "The guest type is unknown";
    }

    // exceptions (class side)

    public static void signalNoPrices() {
        throw new RuntimeException(Floor.pricesListCannotBeEmptyErrorDescription());
    }

    public static void signalNumberOfRoomsMustBeInteger() {
        throw new RuntimeException(Floor.numberOfRoomsMustBeIntegerErrorDescription());
    }

    public static void signalNumberOfRoomsMustBePositive() {
        throw new RuntimeException(Floor.numberOfRoomsMustBePositiveErrorDescription());
    }

    public static void signalPriceMustBeInteger() {
        throw new RuntimeException(Floor.priceMustBeIntegerErrorDescription());
    }

    public static void signalPriceMustBePositive() {
        throw new RuntimeException(Floor.priceMustBePositiveErrorDescription());
    }

    public static void signalRoomNumberDoesNotExist() {
        throw new RuntimeException(Floor.roomNumberDoesNotExistErrorDescription());
    }

    public static void signalUnknownGuestType() {
        throw new RuntimeException(Floor.unknownGuestTypeErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2024-1c-parcial1/20260921-205145-86356/analysis/acceptance-project/src/simshotels/floor.py
from simshotels.room import Room


class Floor:

    # instance creation (class side)

    @classmethod
    def with_rooms_and_prices(cls, a_number_of_rooms, a_price_list):
        cls.assert_is_valid_number_of_rooms(a_number_of_rooms)
        cls.assert_is_valid_price_list(a_price_list)

        return cls(a_number_of_rooms, a_price_list)

    # initialization

    def __init__(self, a_number_of_rooms, a_price_list):
        self._rooms = [Room() for _ in range(a_number_of_rooms)]
        self._prices = a_price_list

    # testing

    def is_available(self):
        return self.total_rooms() == self.total_rooms_available()

    # guests

    def receive_at_room(self, a_guest_type, a_room_number):
        self.room_at(a_room_number).receive(a_guest_type)

    def receive_with_reservation_at_room(self, a_guest_type, a_room_number):
        self.room_at(a_room_number).receive_with_reservation(a_guest_type)

    def reserve_room(self, a_room_number):
        self.room_at(a_room_number).reserve()

    # private

    def room_at(self, a_room_number):
        if a_room_number < 1 or a_room_number > self.total_rooms():
            Floor.signal_room_number_does_not_exist()

        return self._rooms[a_room_number - 1]

    # accounting

    def total_losses(self):
        return self.total_of(lambda room: room.loss_using_if_absent_guest_type(
            self._prices, lambda: Floor.signal_unknown_guest_type()))

    def total_profits(self):
        return self.total_of(lambda room: room.profit_using_if_absent_guest_type(
            self._prices, lambda: Floor.signal_unknown_guest_type()))

    # totals

    def total_rooms(self):
        return len(self._rooms)

    def total_rooms_available(self):
        return self.total_rooms_satisfying(lambda room: room.is_available())

    def total_rooms_occupied(self):
        return self.total_rooms_satisfying(lambda room: room.is_occupied())

    def total_rooms_reserved(self):
        return self.total_rooms_satisfying(lambda room: room.is_reserved())

    def total_rooms_satisfying(self, a_condition):
        return self.total_of(lambda room: 1 if a_condition(room) else 0)

    def total_of(self, a_room_amount):
        return sum(map(a_room_amount, self._rooms))

    # assertions (class side)

    @staticmethod
    def assert_is_integer_if_false(a_number, signals_block):
        if not (isinstance(a_number, int) and not isinstance(a_number, bool)):
            signals_block()

    @staticmethod
    def assert_is_positive_if_false(a_number, signals_block):
        if not (a_number > 0):
            signals_block()

    @staticmethod
    def assert_is_valid_number_of_rooms(a_number_of_rooms):
        Floor.assert_is_positive_if_false(a_number_of_rooms, lambda: Floor.signal_number_of_rooms_must_be_positive())
        Floor.assert_is_integer_if_false(a_number_of_rooms, lambda: Floor.signal_number_of_rooms_must_be_integer())

    @staticmethod
    def assert_is_valid_price_list(a_price_list):
        if len(a_price_list) == 0:
            Floor.signal_no_prices()

        for price in a_price_list.values():
            Floor.assert_is_positive_if_false(price, lambda: Floor.signal_price_must_be_positive())

    # error messages (class side)

    @staticmethod
    def number_of_rooms_must_be_integer_error_description():
        return "Number of rooms must be integer"

    @staticmethod
    def number_of_rooms_must_be_positive_error_description():
        return "The number of rooms must be positive"

    @staticmethod
    def price_must_be_integer_error_description():
        return "Price must be integer"

    @staticmethod
    def price_must_be_positive_error_description():
        return "Price must be positive"

    @staticmethod
    def prices_list_cannot_be_empty_error_description():
        return "Prices List must not be empty"

    @staticmethod
    def room_number_does_not_exist_error_description():
        return "Room number does not exist"

    @staticmethod
    def unknown_guest_type_error_description():
        return "The guest type is unknown"

    # exceptions (class side)

    @staticmethod
    def signal_no_prices():
        raise RuntimeError(Floor.prices_list_cannot_be_empty_error_description())

    @staticmethod
    def signal_number_of_rooms_must_be_integer():
        raise RuntimeError(Floor.number_of_rooms_must_be_integer_error_description())

    @staticmethod
    def signal_number_of_rooms_must_be_positive():
        raise RuntimeError(Floor.number_of_rooms_must_be_positive_error_description())

    @staticmethod
    def signal_price_must_be_integer():
        raise RuntimeError(Floor.price_must_be_integer_error_description())

    @staticmethod
    def signal_price_must_be_positive():
        raise RuntimeError(Floor.price_must_be_positive_error_description())

    @staticmethod
    def signal_room_number_does_not_exist():
        raise RuntimeError(Floor.room_number_does_not_exist_error_description())

    @staticmethod
    def signal_unknown_guest_type():
        raise RuntimeError(Floor.unknown_guest_type_error_description())
